File: electron/backend/modules/audit/audit.service.ts
/**
 * Audit Service
 * Contains business logic for audit operations
 */
import * as auditRepository from './audit.repository.js';
import { AuditLogFilters } from './audit.repository.js';

/**
 * Log an action to the audit trail
 * @param {number} userId - User who performed the action
 * @param {string} action - Action type (create, update, delete, login, logout)
 * @param {string} tableName - Affected table (optional for login/logout)
 * @param {number} rowId - Affected row ID (optional for login/logout)
 * @param {Object} oldData - Previous data (for updates/deletes)
 * @param {Object} newData - New data (for creates/updates)
 * @param {Object} eventDetails - Additional event context
 * @returns {Object} Insert result
 */
export function logAction(
    userId: number,
    action: string,
    tableName: string | null = null,
    rowId: number | null = null,
    oldData: any | null = null,
    newData: any | null = null,
    eventDetails: any | null = null
) {
    return auditRepository.insertAuditLog({
        user_id: userId,
        action,
        table_name: tableName,
        row_id: rowId,
        old_data: oldData,
        new_data: newData,
        event_details: eventDetails
    });
}

/**
 * Log a login event
 * @param {number} userId - User who logged in
 * @param {Object} details - Login details (IP, user agent, etc.)
 */
export function logLogin(userId: number, details: any = {}) {
    return logAction(userId, 'login', null, null, null, null, details);
}

/**
 * Log a logout event
 * @param {number} userId - User who logged out
 * @param {Object} details - Logout details
 */
export function logLogout(userId: number, details: any = {}) {
    return logAction(userId, 'logout', null, null, null, null, details);
}

/**
 * Log a create operation
 * @param {number} userId - User who created the record
 * @param {string} tableName - Table name
 * @param {number} rowId - Created row ID
 * @param {Object} newData - Created data
 */
export function logCreate(userId: number, tableName: string, rowId: number, newData: any) {
    return logAction(userId, 'create', tableName, rowId, null, newData);
}

/**
 * Log an update operation
 * @param {number} userId - User who updated the record
 * @param {string} tableName - Table name
 * @param {number} rowId - Updated row ID
 * @param {Object} oldData - Previous data
 * @param {Object} newData - Updated data
 */
export function logUpdate(userId: number, tableName: string, rowId: number, oldData: any, newData: any) {
    return logAction(userId, 'update', tableName, rowId, oldData, newData);
}

/**
 * Log a delete operation
 * @param {number} userId - User who deleted the record
 * @param {string} tableName - Table name
 * @param {number} rowId - Deleted row ID
 * @param {Object} oldData - Deleted data
 */
export function logDelete(userId: number, tableName: string, rowId: number, oldData: any) {
    return logAction(userId, 'delete', tableName, rowId, oldData, null);
}

/**
 * Get audit logs
 * @param {Object} params - Filter params
 * @returns {Object} Audit logs with pagination
 */
export function getAuditLogs(params: AuditLogFilters) {
    return auditRepository.findAuditLogs(params);
}

File: src/main/java/com/possum/application/audit/AuditService.java
package com.possum.application.audit;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.possum.domain.model.AuditLog;
import com.possum.domain.repositories.AuditRepository;
import com.possum.shared.dto.AuditLogFilter;
import com.possum.shared.dto.PagedResult;

import java.util.List;
import java.util.Objects;

public final class AuditService {

    private final AuditRepository auditRepository;
    private final ObjectMapper objectMapper;

    public AuditService(AuditRepository auditRepository, ObjectMapper objectMapper) {
        this.auditRepository = Objects.requireNonNull(auditRepository, "auditRepository must not be null");
        this.objectMapper = Objects.requireNonNull(objectMapper, "objectMapper must not be null");
    }

    /**
     * Record a generic audit event
     */
    public long recordEvent(Long userId, String action, String tableName, Long rowId,
                            Object oldData, Object newData, Object eventDetails) {
        Objects.requireNonNull(userId, "userId must not be null");
        Objects.requireNonNull(action, "action must not be null");

        try {
            AuditLog auditLog = new AuditLog(
                    null,
                    userId,
                    action,
                    tableName,
                    rowId,
                    toJson(oldData),
                    toJson(newData),
                    toJson(eventDetails),
                    null,
                    null
            );
            return auditRepository.insertAuditLog(auditLog);
        } catch (Exception e) {
            // Audit failures should not break business operations
            System.err.println("Failed to record audit event: " + e.getMessage());
            return -1;
        }
    }

    /**
     * Record a login event
     */
    public long logLogin(Long userId, Object eventDetails) {
        return recordEvent(userId, "login", null, null, null, null, eventDetails);
    }

    /**
     * Record a logout event
     */
    public long logLogout(Long userId, Object eventDetails) {
        return recordEvent(userId, "logout", null, null, null, null, eventDetails);
    }

    /**
     * Record a create operation
     */
    public long logCreate(Long userId, String tableName, Long rowId, Object newData) {
        return recordEvent(userId, "create", tableName, rowId, null, newData, null);
    }

    /**
     * Record an update operation
     */
    public long logUpdate(Long userId, String tableName, Long rowId, Object oldData, Object newData) {
        return recordEvent(userId, "update", tableName, rowId, oldData, newData, null);
    }

    /**
     * Record an update operation with event details
     */
    public long logUpdate(Long userId, String tableName, Long rowId, Object oldData, Object newData, Object eventDetails) {
        return recordEvent(userId, "update", tableName, rowId, oldData, newData, eventDetails);
    }

    /**
     * Record a delete operation
     */
    public long logDelete(Long userId, String tableName, Long rowId, Object oldData) {
        return recordEvent(userId, "delete", tableName, rowId, oldData, null, null);
    }

    /**
     * Get a single audit event by ID
     */
    public AuditLog getAuditEvent(Long auditLogId) {
        Objects.requireNonNull(auditLogId, "auditLogId must not be null");
        return auditRepository.findAuditLogById(auditLogId);
    }

    /**
     * List audit events with filtering and pagination
     */
    public PagedResult<AuditLog> listAuditEvents(AuditLogFilter filter) {
        Objects.requireNonNull(filter, "filter must not be null");
        return auditRepository.findAuditLogs(filter);
    }

    /**
     * List audit events by user
     */
    public List<AuditLog> listAuditEventsByUser(Long userId, int limit) {
        Objects.requireNonNull(userId, "userId must not be null");
        
        AuditLogFilter filter = new AuditLogFilter(
                null, null, userId, null, null, null, null,
                "created_at", "DESC", 1, limit
        );
        return auditRepository.findAuditLogs(filter).items();
    }

    private String toJson(Object obj) {
        if (obj == null) {
            return null;
        }
        if (obj instanceof String str) {
            return str;
        }
        try {
            return objectMapper.writeValueAsString(obj);
        } catch (JsonProcessingException e) {
            return "{\"error\": \"Failed to serialize data\"}";
        }
    }
}
